### ISSUE TITLE
Differing types of a singular polymorphic type do not seem to be checked.

### ISSUE THREAD

**Issue by paulcoyle** [facebook/flow#2016]
[You can see this in action here.](https://github.com/paulcoyle/flow-react-polymorphic)

I am unsure if this is a problem with flow, it's relationship to react, or my syntax, but I would expect the following to produce an error indicating that there is a type inconsistency between the values for the `alpha` and `beta` props:

`Main.jsx`

``` javascript
/* @flow */
import React from 'react'
import Consistency from './Consistency'

export class Main extends React.Component {
  render() {
    return <Consistency alpha={1} beta="2" />
  }
}

export default Main
```

`Consistency.jsx`

``` javascript
/* @flow */
import React from 'react'

type Props<T> = {
  alpha: T;
  beta: T;
}

export class Consistency<T> extends React.Component<void, Props<T>, void> {
  props: Props<T>;

  render() {
    return <div>{this.props.alpha} : {this.props.beta}</div>
  }
}

export default Consistency
```

My thinking is that the polymorphic type, `T`, of `Consistency` would be inferred by the given props and, if they are not all of the same type, this is a type error.  Of course, this would also extend to polymorphic default props and state types, e.g.

``` javascript
type DefaultProps<T> = {
  alpha: T;
}

type Props<T> = DefaultProps<T> & {
  beta: T;
}

type State<T> = {
  stateAlpha: T;
  stateBeta: T;
}

class Everything<T> extends React.Component<DefaultProps<T>, Props<T>, State<T>> {
   // etc.
}
```

**Comment by asolove** [comment]
This is very confusing, but I think Flow is ultimately doing the right thing, just explaining it very poorly.

To reproduce the issue, I lumped the code into one file and then made some minor syntactic changes to upgrade to Flow v0.53.1. [Here's the result](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA0wCeYWcACsWKgDwAqAfHAC8cAN404KADZgAFsgBccHgG5xcAEZYYCpaoC+dLAA9IsOLklp0AYQgA7VMFQwsd3A14DjLuwBN02HgwAHS24PauMFzsEJyeADRwAG4QwL4CYhKUflhQABQAlKJqWVoArlB2cFy+wEl8IjAyTsFgHKjByNJyenCKjc0dbbEdmtp6XAD0tfVqBgY0FlZwALLIwFXerv6YVCFhkHaRXClpiafpxaU5+UWZEnCUMBVVXLYOTj7uUrLIggBEAEZ-hotH9-gAmEGTPhzGgGIA)

The error Flow gives is:

```
10:     return <div>{this.props.alpha} : {this.props.beta}</div>
                     ^ T. This type is incompatible with
[LIB] static/v0.53.1/flowlib/react.js:23:   | Iterable<React$Node>;
                                                       ^ union: undefined | null | boolean | number | string | type application of type `React$Element` | type application of identifier `Iterable`
```

I honestly had no idea what this meant for a while. I played around with the code a lot to try to understand what the heck was trying to unify with `Iterable<React$Node>`. 

Finally, it occurred to me: within JSX, there are rules about what can be output. For example, you can't have an expression like

```
let plus1 = (x) => x+1;
<div>{plus1}</div>
```

Because functions aren't allowed as JSX content, how would they be printed? Sure enough, when I tried that example, I got a very similar type error:

```
21: <div>{plus1}</div>;
          ^ function. This type is incompatible with
[LIB] static/v0.53.1/flowlib/react.js:461:       children?: React$Node,
                                                            ^ union: undefined | null | boolean | number | string | type application of type `React$Element` | type application of identifier `Iterable`
```

So this error is what we get when we try to stick something in JSX output that React doesn't know what to do about. React can deal with that lost of alternatives: undefined, null, boolean, etc... but other types won't work!

And sure enough, if we look back at the original example, we are putting into JSX the values `this.props.alpha`, when all we know about that value is that it has type `T`. It could be anything! So Flow is complaining, saying: hey, you are trying to put anything in JSX, but I only know how to use a specific set of things there.

When we run into an issue like this, usually the solution is to put a restriction on the type variable, so that instead of being any time, it can instead be any of a set of types that meet some criteria. In our case, that criteria is "things JSX can understand", which happens to be the type `React.Node` mentioned in the error message.

Sure enough, if we change our class declaration to:

```js
export class Consistency<T: React.Node> extends React.Component<Props<T>, void> {
```

then [the resulting code](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA0wCeYWcACsWKgDwAqAfHAC8cAN404KADZgAFsgBccHgG5xcAEZYYCpaoC+dLAA9IsOLklp0AYQgA7VMFQwsd3A16LseGADoAchAAJlgCxi52Qeje+L624PauMFzsEJy8fAA0cABuEMBBAmISlJFYUAAUAJSiaiVaAK5QdnBcQcA5fCIwMk6+YByovsjScnpwit29QwNpQ5raelwA9O2dagYGNBZWcACyyMAt4a5RmFR+8ZB2SVx5Bdn3hbX1ZZU1xRJwlDBNLVy2BxOCLuKSyZCCABEAEZIRotBDIQAmOHLPgbGh6IA) type checks just fine.

Hope that explanation makes sense! I'm happy to discuss more, but just wanted to make clear this was a usage/error message issue rather than a random bug in Flow.

I believe we can close this issue (/cc @calebmer)

**Comment by johnryan** [comment]
@asolove seeing a similar issue to this after upgrading flow. See [this example](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQdWAHpLHDAJ5hZwAiEAczgBeOAEEAdqGQAbOADI4Abzpw4BAK5QAwhBnQAXHFQwowCQIA0dAL71GLaPE7dxUkLJHLVcCchBYRiZmFrb2LjwACsRg6KIqatwwRgD8-EK2DJoS+MAQEnCRWDC64HAAFGAxqEbRELEAlN5quPkmynBJcDZeVfWo9ACQlDBaBQA8ACbAAG4AfEpd8opJAHR+ATbjAPTT85l02bn5cADq0ADW5gJFJSRgFX2xtdVNCXCtEu1KncXdvdUhiMxnAprMFpU-gBCUQSDQyGRNZa-GDrfxYLa7cEHca3UoPJLCJTvDaBOAAcgAQhAAEbk6xqTQ6PSGCk04gAdwk5NUNhs2zmdCAA)

It seems to me like both these usages should be valid. Can you shed any light on this? If it's a different issue i'm happy to open a new ticket.

**Comment by asolove** [comment]
Doing intersections of inexact object types doesn’t work the way you might expect.  If you use type spread, this works and corresponds to the intent of the code:

```
export type Dog = {
  ...Animal,
  furColor: string,
};
```

**Comment by johnryan** [comment]
@asolove Thanks for the explanation, I definitely prefer that syntax. However that seems to cause a different issue [here](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQdWAHpLHDAJ5hZwAiEAczgBeOAG86cOADpZAQQB2oZABsANJLgEArlADCEFdABccVDCjAFAjQF96jFtHiducAHIQFAZTCVkACb8QqISUgrIIFim5pbWGlI6+oYmZhZWNnT2DMysLlw8isoqIuKaEVEx6dZZDq48AArEYOihmtwwpgD8wXBZDPUeXr7+AU0QLaVhcB3dnj5+OEGCfbYMdDoK+MBeHpFYBuAAFH4TqKZiAG6qVXECtgCU05QwugpwADwBwJcAfGKnFrSa4qWwfAD03z+-Q22i2MB27waWBghzAcBOzXOcHGLQeZSkuC85nEMxRcFspUBqHoAEgXm9PlD-h04AAyNmfdz7NFwEHCAEo6QVLC2cG-MGQn6-GF0cHgrTQMwkHj+VC7MgishwABGWCJUXQyHeEwRXlUfNU2h4AHcABZYd5gNCoLABOBWPnAZBmRaBMwcBQwZBMOUKgCMpht0AA1ugdRw4NAAOQZMkwYX7OA24AwO1wZPJ2HwxFwADqsYyyNRJDA4Yx1NMuNQ+OmRIUJLE6YpVKxdIZUHeX2lgvgHK5PNrlpUAo6maicAAPouC8mxRKIcyYfK4JHs7H44ntKg09Ts7m7RBtPBUH73ahA8HQ5ttrsK1AY1WUWiAEwNrGmPMIxLM2raaO2nbdpSojUv2KKMsOfyjuynIfNyUS8vyo7zqK4qSluax0B81a8h0ArTCKpjJgAQhAOrJgkWi6AYRhQFROrEDaChFhS650EAA)

Basically, the props that come in via spread syntax end up being evaluated as optional (as far as i can tell).
